### PR TITLE
First version of code and style guidelines for cpp.

### PR DESCRIPTION
**Pull request by jlingema**
Hello all,

This is the first draft for both style and code guidelines for C++, as shown in the meeting on 25.11. The paragraph regarding package structure was extended to what was discussed.

Regarding the style-guide: I broadly follow the LHCb (= Gaudi) style guide, a link is in the file below.

Unfortunately, github has no pull-requests for the wiki pages. I put the guides in .md files in `FCCSW/doc`. A final place for them should probably be discussed as well.

I'm happy to incorporate your comments.

Cheers,
Joschka

PS: The files are probably easier to read in my fork:
[general guidelines](https://github.com/jlingema/FCCSW/blob/master/doc/GeneralCppGuidelines.md) and [style guide](https://github.com/jlingema/FCCSW/blob/master/doc/CppCodingStyleGuidelines.md)

### CODE REVIEW

File: doc/GeneralCppGuidelines.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+void printSqrt(float value) {
+  cout << sqrt(value) << endl;
+}
+```
+
+## Pointers and References
+The code should not contain `new` outside of braces. Avoid raw-pointers where possible.
+
+Raw C++ pointers should be avoided. Only use them when you are certain that you have to. Consider using `std::unique_ptr`, `std::shared_ptr` and references instead.
```
Comment (hegner) [inline doc/GeneralCppGuidelines.md]: Maybe some code example would be nice.

File: doc/GeneralCppGuidelines.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+### Re-entrancy
+If a function cannot be stateless, the state should be local.
+
+_currently a stub_
+
+## Useful C++11 Features
+### Auto
+Where the type of an object can be inferred by the compiler, `auto` can be used.
+
+It can be useful to decouple from specific implementations of data-structures. If constness is ambiguous, you can use `const auto`. Similarly you can make sure to get a reference with `const auto&`.
```
Comment (hegner) [inline doc/GeneralCppGuidelines.md]: in fact "auto" just doesn't include constness... that's an orthogonal concept.

File: doc/GeneralCppGuidelines.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+## Other Things to Avoid
+- Do not write singletons
+- Do not define global variables, any entity in the global namespace
+- Never have non-const static functions or member data
+
+### Avoid C syntax
+While most C-code compiles as C++, it is strongly discouraged.
+
+An incomplete list:
+- Use const-expressions instead of macros (see below for details)
```
Comment (hegner) [inline doc/GeneralCppGuidelines.md]: there are no details below ;-)

File: doc/CppCodingStyleGuidelines.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+Consider additional `.impl` files for performance critical `inline` functions that are longer than one line or templated functions.
+
+*Note*: Functions should only be inlined if they are proven to help the compiler to optimise the code.
+
+***
+
+## Function Declaration
+All entities should be contained in a namespace.
+### Parameter Order
+Parameter order is: constant inputs, editable inputs and then outputs (if any).
```
Comment (hegner) [inline doc/CppCodingStyleGuidelines.md]: constant inputs, non-const inputs, outputs.

File: doc/GeneralCppGuidelines.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+All entities are contained in a namespace.
+
+_Warning_: Do not include header files within the namespace region but before!
+
+### Non-member Functions and Namespaces
+If non-member functions are associated to a specific class they belong in the same namespace as the class.
+
+Non-member functions that are not associated to specific classes are grouped in namespaces by functionality.
+
+### Avoid Using Namespace
```
Comment (zaborowska) [inline doc/GeneralCppGuidelines.md]: -> Avoid Using "Using Namespace"

File: doc/GeneralCppGuidelines.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+private:
+ Vector3 m_position;
+};
+```
+
+## Lifetime
+Always consider what the lifetime of an entity is and restrict it to the shortest period that makes sense.
+
+Sometimes it is not inherently clear what the lifetime of an object in memory is. While names only exist in the scope in which they are defined, the objects they refer to can exist longer:
+- Objects created with `new` live until they are `delete`d
```
Comment (zaborowska) [inline doc/GeneralCppGuidelines.md]: a typo (`delete`d -> `deleted`)

File: doc/CppCodingStyleGuidelines.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+* [Code Formatting](#code-formatting)
+   * [Indentation](#indentation)
+   * [Curly Braces](#curly-braces)
+   * [Comments](#comments)
+
+## Naming Conventions
+### Variable Names
+Names should be as descriptive as possible.
+
+Do not worry about space; Worry about how understandable the name is. Avoid unclear abbriviations.
```
Comment (zaborowska) [inline doc/CppCodingStyleGuidelines.md]: a typo: abbreviations

File: doc/GeneralCppGuidelines.md (hunk 6 above)
Comment (jlingema) [inline doc/GeneralCppGuidelines.md]: I'll rephrase it to "until destroyed using `delete`". I wanted to have the keyword as inline code.

**Comment by hegner** [comment]
Is that still WIP or are you happy to share this?

**Comment by jlingema** [comment]
I think as a first version it is fine. Some things (like further reading) can still be extended later.
